sympy/matrices/decompositions.py

```diff
@@ -1343,3 +1344,72 @@ def _QRdecomposition(M):
     QRsolve
     """
     return _QRdecomposition_optional(M, normalize=True)
+
+def _upper_hessenberg_decomposition(A):
+    """Converts a matrix into Hessenberg matrix H
+
+    Returns 2 matrices H, P s.t.
+    $P H P^{T} = A$, where H is an upper hessenberg matrix
+    and P is an orthogonal matrix
+
+    Examples
+    ========
+
+    >>> from sympy import Matrix
+    >>> A = Matrix([
+    ...     [1,2,3],
+    ...     [-3,5,6],
+    ...     [4,-8,9],
+    ... ])
+    >>> H, P = A.upper_hessenberg_decomposition()
+    >>> H
+    Matrix([
+    [1,    6/5,    17/5],
+    [5, 213/25, -134/25],
+    [0, 216/25,  137/25]])
+    >>> P
+    Matrix([
+    [1,    0,   0],
+    [0, -3/5, 4/5],
+    [0,  4/5, 3/5]])
+    >>> P * H * P.H == A
+    True
+
+
+    References
+    ==========
+
+    .. [#] https://mathworld.wolfram.com/HessenbergDecomposition.html
+    """
+
+    M = A.as_mutable()
+
+    if not M.is_square:
+        raise NonSquareMatrixError("Matrix must be square.")
+
+    if M.is_upper_hessenberg:
+        return M, M.eye(M.cols)
+
+    n = M.cols
+    P = M.eye(n)
+    H = M
+
+    for j in range(n - 2):
+
+        u = H[j + 1:, j]
+
+        if u[1:, :] == M.zeros(1, u.rows - 1):
```

Comment:
Why not use `is_zero_matrix`?

Reply:
I didn't know about it, I was searching for something like that. I'll change it thanks
Thanks